### PR TITLE
Crafting refactor

### PR DESCRIPTION
**Pull request by mugling**
The crafting and recipe code is horrible:
- Load order is byzantine and validation occurs at multiple sites in the code
- When profiled `recipe_dictionary` is inefficient due to usage of `std::list` and pointer indirection
- JSON recipes involve excessive and error-prone duplication
- Extensive use of pointers and negligible use of `const`

Refactoring:
- [x] Split `recipe` from `uncraft`
- [X] Support `strict` mode parsing (needs JSON cleanup in part 2)
- [x] Support `copy-from` syntax for `recipe`
- [X] Move all checks for `recipe` validity to `recipe_dictionary::finalize`
- [X] Refactor `recipe_dictionary` to `std::map`
- [X] Refactor `recipe *` to `recipe &` (mostly)

### CODE REVIEW

File: data/json/recipes/recipe_medsandchemicals.json [hunk 1]
```diff
@@ -2365,7 +2365,7 @@
 },{
   "type" : "recipe",
   "result": "slime_scrap",
-  "byproducts": "slime_scrap",
+  "byproducts": [ [ "slime_scrap" ] ],
```
Comment (Coolthulhu) [inline data/json/recipes/recipe_medsandchemicals.json]: This one could probably be changed into a multiplier.
We could also have a warning on byproduct being the same as result.

File: src/crafting.cpp [hunk 2]
```diff
@@ -42,268 +43,134 @@ recipe::recipe() :
...
+                r.learn_by_disassembly[skill_id( arr.get_string( 0 ) )] = arr.get_int( 1 );
             }
-            t->book->recipes.insert( rwd );
         }
-        r->booksets.clear();
+    }
 
-        if( !item::type_is_defined( r->result ) ) {
-            buffer << "Recipe " << r->ident() << " defines invalid result " << r->result << "\n";
+    if( jo.has_member( "byproducts" ) ) {
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Could be a good idea to special case the most common case: single array.
First half of this PR is just adding extra brackets to everything.

File: src/crafting.cpp [hunk 3]
```diff
@@ -42,268 +43,134 @@ recipe::recipe() :
...
-                buffer << "book " << book_id << " for recipe " << r->ident() << " does not exist" << "\n";
-                continue;
-            }
-            const itype *t = item::find_type( book_id );
-            if( !t->book ) {
-                // TODO: we could make up a book slot?
-                buffer << "book " << book_id << " for recipe " << r->ident() << " is not a book" << "\n";
-                continue;
+        if( jo.has_int( "decomp_learn" ) ) {
+            if( !r.skill_used ) {
```
Comment (Coolthulhu) [inline src/crafting.cpp]: What is the `!` operator? The usage is ambiguous here - does it mean that the skill is set or is valid?
Check for validity should be in verification/finalization, to avoid needless order dependence.

File: src/crafting.cpp (hunk 2 above)
Comment (mugling) [inline src/crafting.cpp]: Prior to this PR there are four possible forms `byproducts` (although not all were used). It's very difficult to write automated tooling that doesn't struggle with that (there is a tendency to flip between two semantically equivalent forms).

If we are going to aim for better tooling (linters, editors etc) and still want to maintain ability to edit JSON directly (definitely important) then having a canonical form is important.

File: data/json/recipes/recipe_medsandchemicals.json (hunk 1 above)
Comment (mugling) [inline data/json/recipes/recipe_medsandchemicals.json]: Noted. I'll add that warning.

File: src/crafting.cpp [hunk 4]
```diff
@@ -1831,33 +1689,7 @@ void player::complete_disassemble( int item_pos, const tripoint &loc,
 
 const recipe *recipe_by_name( const std::string &name )
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Would be nice to constify this and change the pointer to reference if possible.

File: src/crafting.h [hunk 5]
```diff
@@ -59,16 +47,21 @@ struct recipe {
...
         std::string cat;
         // Does the item spawn contained in container?
-        bool contained;
+        bool contained = false;
         // What does the item spawn contained in? Unset ("null") means default container.
-        itype_id container;
+        itype_id container = "null";
         std::string subcat;
-        skill_id skill_used;
+        skill_id skill_used = skill_id::NULL_ID;
```
Comment (Coolthulhu) [inline src/crafting.h]: `NULL_ID` in this context may not be safe.

File: src/crafting.cpp (hunk 4 above)
Comment (mugling) [inline src/crafting.cpp]: Indeed one of the goals is to get rid of the pointer shenanigans. Would you prefer I attempted for this in a single PR?

File: src/iuse.cpp [hunk 6]
```diff
@@ -5964,20 +5964,19 @@ bool einkpc_download_memory_card(player *p, item *eink, item *mc)
...
+                if( r.valid_learn() ) {
+                    if (r.difficulty >= 3 && one_in(r.difficulty + 1)) {
+                        candidates.push_back( &r );
                     }
                 }
             } else {
-                if( ( elem )->cat == "CC_FOOD" ) {
-                    if (dif <= 3 && one_in(dif)) {
-                        candidates.push_back( elem );
+                if( r.cat == "CC_FOOD" ) {
```
Comment (Coolthulhu) [inline src/iuse.cpp]: While not introduced here, this is a horrible, horrible hack.

File: src/crafting.h (hunk 5 above)
Comment (mugling) [inline src/crafting.h]: Suggestions?

Comment (Coolthulhu) [inline src/crafting.h]: The same hack as in other places: wrapping it in an explicit constructor with null value.

**Comment by mugling** [comment]
@Coolthulhu updated PR preamble to include outline of intended work

File: src/crafting.cpp (hunk 3 above)
Comment (mugling) [inline src/crafting.cpp]: Load order isn't relevant here - the validity of `skill_used` is checked in `finalize`. This just checks that `skill_used` is non-empty as it is an error to specify eg. `"decomp_learn": 3` without `skill_used`

File: src/crafting.h [hunk 7]
```diff
@@ -24,29 +24,19 @@ typedef int nc_color; // From color.h
...
 
     public:
-        itype_id result;
-        int time; // in movement points (100 per turn)
-        int difficulty;
-        bool valid_to_learn;
+        recipe() : skill_used( NULL_ID ) {}
+
+        itype_id result = "null";
+        int time = 0; // in movement points (100 per turn)
```
Comment (codemime) [inline src/crafting.h]: `moves` or `time_in_moves` or `moves_to_perform` would make the comment redundant and also would eliminate confusion in the implementation.

File: src/crafting_gui.cpp [hunk 8]
```diff
@@ -625,25 +625,14 @@ int recipe::print_items( WINDOW *w, int ypos, int xpos, nc_color col, int batch
     const int oldy = ypos;
 
     mvwprintz( w, ypos++, xpos, col, _( "Byproducts:" ) );
-    for( auto &bp : byproducts ) {
-        print_item( w, ypos++, xpos, col, bp, batch );
+    for( const auto &bp : byproducts ) {
+        item it( bp.first, calendar::turn, item::default_charges_tag{} );
+        mvwprintz( w, ypos++, xpos, col, _( "> %d %s" ), bp.second * batch, it.tname().c_str() );
```
Comment (codemime) [inline src/crafting_gui.cpp]: Can `item::display_name()` be used here? (to prevent constructing full blown object above).

File: src/item.cpp [hunk 9]
```diff
@@ -1835,17 +1835,15 @@ std::string item::info( bool showtext, std::vector<iteminfo> &info ) const
         } else { // use the contained item
             tid = contents.front().typeId();
         }
-        const std::vector<recipe *> &rec = recipe_dict.of_component( tid );
+        const auto& rec = recipe_dict.of_component( tid );
```
Comment (codemime) [inline src/item.cpp]: `const auto &rec` would be more consistent.

File: src/iuse.cpp [hunk 10]
```diff
@@ -5964,20 +5964,19 @@ bool einkpc_download_memory_card(player *p, item *eink, item *mc)
...
+        for( const auto &e : recipe_dict ) {
+            const auto r = e.second;
 
-            const int dif = ( elem )->difficulty;
-
-            if (science) {
-                if( elem->valid_learn() ) {
-                    if (dif >= 3 && one_in(dif + 1)) {
-                        candidates.push_back( elem );
+                if (science) {
```
Comment (codemime) [inline src/iuse.cpp]: Wrong indentation.

File: src/iuse.cpp [hunk 11]
```diff
@@ -5964,20 +5964,19 @@ bool einkpc_download_memory_card(player *p, item *eink, item *mc)
...
+                if( r.valid_learn() ) {
+                    if (r.difficulty >= 3 && one_in(r.difficulty + 1)) {
```
Comment (codemime) [inline src/iuse.cpp]:
```c++
if( r.valid_learn() && r.difficulty >= 3 && one_in(r.difficulty + 1))
```

File: src/recipe_dictionary.cpp [hunk 12]
```diff
@@ -1,95 +1,289 @@
...
+            continue;
+        }
+
+        if( std::any_of( r.byproducts.begin(),
+        r.byproducts.end(), []( const std::pair<itype_id, int> &bp ) {
+        return !item::type_is_defined( bp.first );
+        } ) ) {
+            debugmsg( "Recipe %s defines invalid byproducts", id );
+            it = recipe_dict.recipes.erase( it );
+            continue;
```
Comment (codemime) [inline src/recipe_dictionary.cpp]: It would be nice to extract removal of invalid recipes into `std::remove_if()`. It would prevent this ugly repetition of 
```c++
it = recipe_dict.recipes.erase( it );
continue;
```
and the unpleasant `for( auto it = recipe_dict.recipes.begin(); it != recipe_dict.recipes.end(); ) {`. Also, that would be more structured.

File: src/recipe_dictionary.cpp [hunk 13]
```diff
@@ -1,95 +1,289 @@
...
+    recipe_dict.recipes.clear();
 }
 
-const std::vector<recipe *> &recipe_dictionary::of_component( const itype_id &id )
+void recipe_dictionary::delete_if( const std::function<bool( const recipe & )> &pred )
 {
-    return by_component[id];
+    for( auto it = recipes.begin(); it != recipes.end(); ) {
+        if( pred( it->second ) ) {
+            it = recipe_dict.recipes.erase( it );
```
Comment (codemime) [inline src/recipe_dictionary.cpp]: `std::remove_if()` wouldn't hurt here too.

File: src/recipe_dictionary.cpp (hunk 12 above)
Comment (mugling) [inline src/recipe_dictionary.cpp]: I don't think `std::remove_if` works for `std::map`?

File: src/recipe_dictionary.cpp (hunk 13 above)
Comment (mugling) [inline src/recipe_dictionary.cpp]: As above

File: src/recipe_dictionary.cpp [hunk 14]
```diff
@@ -1,95 +1,289 @@
...
+
+    auto &r = recipe_dict.recipes[ id ];
+    r.result = result;
+    r.ident_ = id;
+
+    if( uncraft ) {
+        r.cat = "CC_NONCRAFT";
+        r.subcat = "CSC_NONCRAFT";
+        r.reversible = true;
+        r.valid_to_learn = false;
```
Comment (codemime) [inline src/recipe_dictionary.cpp]: Some kind of `recipe::load()` would be useful here. All these `r.whatever` clearly signals that they want to be in the `recipe` class. They would also acquire some intimacy there (being private if needed).

Comment (mugling) [inline src/recipe_dictionary.cpp]: Yes that is the long-term goal but possibly not in this PR.

File: src/recipe_dictionary.cpp (hunk 12 above)
Comment (codemime) [inline src/recipe_dictionary.cpp]: Fair point, it won't work, but a custom template that implements the idiom would (and it would certainly be useful elsewhere). I think you would agree that duplicated toying with the iterator in a dozen of clauses is quite fragile and rather ugly.

Comment (mugling) [inline src/recipe_dictionary.cpp]: I almost wrote a macro...

File: src/crafting_gui.cpp (hunk 8 above)
Comment (mugling) [inline src/crafting_gui.cpp]: You mean `item::nname`?

File: src/iuse.cpp (hunk 11 above)
Comment (mugling) [inline src/iuse.cpp]: Not semantically equivalent and given that code is a nasty hack I'd rather not make any unrelated changes here

Comment (codemime) [inline src/iuse.cpp]: Why not? Two nested clauses:
```c++
if( a ) {
    if( b ) {
      if( c ) {
          ...
      }
    }
} else {
    ...
}
```

are equivalent to one nested clause:

```c++
if( a ) {
    if( b && c ) {
      ...    
    }
} else {
    ...
}
```

or no nested clauses (preferable):

```c++
if( !a ) {
    ...    
} else if( b && c ) {
    ...
}
```

File: src/crafting_gui.cpp (hunk 8 above)
Comment (codemime) [inline src/crafting_gui.cpp]: Yeah, you're right.

**Comment by mugling** [comment]
Ready. No intended behavioral changes

### Major changes:
* `recipe` and `uncraft` now have separate namespaces
* constraints are applied and strict mode could later be enabled in `recipe_dictionary::load`
* `byproducts` has a single syntax
* All checks now occur at a single site in `recipe_dictionary::finalize`
* Tearing up of books is no-longer special cased via `fake_recipe_book`
* Inheritance via `copy-from` and `abstract` is supported
* Deferred loading of JSON entries is moved to `DynamicDataLoader::load_deferred`

ff95ee3 provides an example of what is possible in terms of inheritance (although we needn't necessarily go with this design when we come to completing #17904).

### Future work (separate PR)
* move `recipe_category` into `recipe_dictionary` and validate the names
* Allow `i18n` of recipe category names
* Further refactor `recipe_dictionary` to function similar to other container singletons
* Extended documentation for `recipe` class members and add GH wiki overview

### Inheritance
* Undefined fields are inherited/defaulted whereas defined fields as overwritten
* Child recipes inherit `using` requirements from their parent (if any)
* Inline requirements are always overwritten
* Mods continue to overwrite `core` content
* `uncraft` has priority over `reversible recipe`
* Books disassembly is special case only if neither `uncraft` or `reversible recipe` exists

File: data/json/uncraft/ammo/9mm.json [hunk 15]
```diff
@@ -2,61 +2,29 @@
...
-    "components": [
-      [ [ "lead", 2 ] ],
-      [ [ "9mm_casing", 1 ] ],
-      [ [ "smpistol_primer", 1 ] ],
-      [ [ "gunpowder", 4 ] ]
-    ],
-    "flags": [ "UNCRAFT_SINGLE_CHARGE" ]
+    "using": [ [ "ammo_9mm", 1 ], [ "ammo_bullet", 1 ] ],
+    "components": [ [ [ "gunpowder", 4 ] ] ],
+    "copy-from": "ammo_cartridge"
```
Comment (Coolthulhu) [inline data/json/uncraft/ammo/9mm.json]: Why is `"copy-from"` on the end of the json in recipes?

Comment (mugling) [inline data/json/uncraft/ammo/9mm.json]: An omission in the linter. I'll fix that now

**Comment by mugling** [comment]
Next merge test of all my open PR's this one most important, tagged as priority

**Comment by Coolthulhu** [comment]
Causes unrecoverable errors with the `using` in reloading recipes.

**Comment by mugling** [comment]
Do you mean it segfaults? How do you know it's `using` that does it?

**Comment by Coolthulhu** [comment]
Doesn't segfault, but goes back to menu.
The problem seems to be that some recipes (the ammo reloading ones) have usings with no values, for example `[ "ammo_9mm" ]` rather than `[ "ammo_9mm", 1 ]`.

**Comment by mugling** [comment]
Fixed

**Comment by Coolthulhu** [comment]
* Longcraft is broken

**Comment by mugling** [comment]
What is longcraft?

**Comment by Coolthulhu** [comment]
Major:
* Recipe searching looks broken

**Comment by Coolthulhu** [comment]
Ok, now this is pretty hilarious:
I just managed to uncraft a rag into 250ml of water and a blood soaked rag. Needed a fire to do that.

**Comment by mugling** [comment]
> Pressing an unassigned button when in "item menu" (the one with highlighted "activate", "eat", "autopickup" etc.) prints "you can't disassemble [item name]" if the item can't be disassembled, does nothing otherwise

Can't reproduce and there are no changes that could explain this. What are the exact steps. May be pre-existing undefined behavior?

**Comment by mugling** [comment]
> I just managed to uncraft a rag into 250ml of water and a blood soaked rag. Needed a fire to do that.

How?

**Comment by Coolthulhu** [comment]
> What is longcraft?

It's in keybinds as "craft as long as possible". It is supposed to restart crafting until it is no longer possible.

> Can't reproduce and there are no changes that could explain this. What are the exact steps.

Check up on item, press an unbound key. For example, 'A'.
> May be pre-existing undefined behavior?

Possibly. It's quite minor.

> How?

Spawned a blindfold, uncrafted it into bandana, uncrafted it into rags. Noticed rags can't be uncrafted just like that, so I checked up on them. Item info said they can be uncrafted into water and bloodied rags. I attempted uncraft, it said I need fire. Spawned fire, uncrafted rag, it worked.

**Comment by Coolthulhu** [comment]
I just got infinite loop trying to "cut up all you can" over a fresh zombie corpse with some clothing.

**Comment by mugling** [comment]
I'm short on time now. If you enumerate the faults and presuming you have no objection I'll fix and push tomorrow?

**Comment by Coolthulhu** [comment]
Major:
* Infinite loop when cutting up all items, possibly requiring an uncuttable item on the stack
* Recipe searching produces no results
* Some items have invalid uncraft recipes selected for them. Notably the rag, which finds "rag from bloodied rag and water" recipe instead of "deconstruct into thread" recipe
* Recraft is broken, does nothing

Minor:
* Just viewing item info prints "You cannot disassemble x" if it can't be disassembled
* Longcraft is broken, only crafts one item

**Comment by mugling** [comment]
That's some fairly epic breakage

**Comment by mugling** [comment]
Think I have it sorted, checking all the deficiencies are fixed

**Comment by mugling** [comment]
Yes, fixed

**Comment by Coolthulhu** [comment]
Here's what I'd test:
* Spawn 10 planks
* Do longcraft of pointy stick. Should craft 10 pointy sticks without prompts

* Spawn 10 planks, drop 5
* Do longcraft of pointy stick. Should prompt, craft 5 sticks, prompt again, craft 5 sticks.

* Spawn 10 planks, 20 leather, 60 rags
* Craft one bindle with leather. Recraft. Should not prompt for ingredients.
* Longcraft bindles with leather. Should not prompt until leather runs out

* Spawn zombie and debug-kill it. Spawn another on top of it and debug kill it.
* Drop 2 flashlights on the corpses
* Go over the corpses and cut up everything. Should not dismantle flashlights

* Repeat spawning corpses and dropping flashlights
* Go over the corpses
* Dismantle everything for as long as possible
* Should not butcher or cut anything, but should dismantle both flashlights

* Turn on autosave and set to some short time
* Prepare some long craft (for example, forge craft). Debug hammerspace must be off.
* Wait (not in game, just for the autosave timer to run out)
* Start crafting some long craft
* After autosave triggers, forcefully close the game
* Reload the save
* Must not crash. Can produce the object for free, consume the resources without producing or whatever, just not crash

**Comment by mugling** [comment]
Yes, that all works fine

**Comment by mugling** [comment]
@Rivet-the-Zombie if your feeling brave this one as well. I've run @Coolthulhu's checks and want to use this new functionality to finish off #17904

**Comment by Rivet-the-Zombie** [comment]
@mugling, I'll have to look into this one tomorrow. The sun is coming up here and it's time for me to crawl back into my grave for the day.

**Comment by mugling** [comment]
No worries. As soon as we can merge #17904 we close another milestone blocker

**Comment by Coolthulhu** [comment]
I'm getting segfaults on start, before I even get to menu.
Either this or the recoil PR. And I don't see anything in the recoil PR that would cause it.
Here you moved a constructor from `.cpp` to `.h`, meaning it may be invoked before `skill_id::NULL_ID` is initialized. That was a cause of a number of known segfaults.

**Comment by mugling** [comment]
Consistent or random?

**Comment by mugling** [comment]
And specifically when. Before the splash screen, on chargen, on world load? Any stack trace?

**Comment by mugling** [comment]
Can't see anything in recoil PR either. Try this branch with latest commit pushing constructor back into implementation

**Comment by Coolthulhu** [comment]
Still can't get it to run, I'll try a debug build.

**Comment by Coolthulhu** [comment]
Got this backtrace:
```
#0  std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (this=0x1012b79f8 <null_recipe+312>,
    __str=<error reading variable: Cannot access memory at address 0xffffffffffffffe8>) at /usr/lib/gcc/x86_64-pc-msys/5.3.0/include/c++/bits/basic_string.tcc:617
#1  0x0000000100aa5959 in string_id (this=0x1012b79f8 <null_recipe+312>) at src/string_id.h:66
#2  recipe::recipe (this=0x1012b78c0 <null_recipe>) at src/crafting.cpp:135
#3  0x0000000100c37e25 in __static_initialization_and_destruction_0 (__initialize_p=<optimized out>, __priority=<optimized out>) at src/recipe_dictionary.cpp:14
#4  0x0000000100c40c95 in _GLOBAL__sub_I_recipe_dict () at src/recipe_dictionary.cpp:401
#5  0x0000000180046245 in do_global_ctors (force=0, in_pfunc=0x101013f40 <___CTOR_LIST__>) at /msys_scripts/msys2-runtime/src/msys2-runtime/winsup/cygwin/dcrt0.cc:78
#6  __main () at /msys_scripts/msys2-runtime/src/msys2-runtime/winsup/cygwin/dcrt0.cc:1192
#7  0x0000000100c42e2c in main (argc=1, argv=0xffffcc20) at src/main.cpp:58
```

Looks like the null recipe in recipe dictionary invokes recipe constructor with `NULL_ID` for skill id, which isn't initialized yet.

**Comment by Coolthulhu** [comment]
OK, I'm out of ideas: I replaced the `NULL_ID` with `"none"` and it still crashes, this time with no valid backtrace.

**Comment by mugling** [comment]
With or without the last commit moving the constructor?

**Comment by mugling** [comment]
Try moving the static `null_recipe` inside the two dependent functions in `recipe_dictionary.cpp`.

```
> __main () at /msys_scripts/msys2-runtime/src/msys2-runtime/winsup/cygwin/dcrt0.cc:1192
```

We've now started targeting cygwin? This is going to force all development through a bottleneck whilst a lot of platform-specific bugs suddenly become live.

**Comment by mugling** [comment]
Bump: edited above - I can't compile that fix locally so can you check

**Comment by Coolthulhu** [comment]
> We've now started targeting cygwin?

Master branch works just fine, it's just this PR that breaks it.
It looks much more like a problem that your compiler has hidden than one that cygwin uncovered.

One of the guides advocates using MSYS2, which supplies cygwin libraries.

**Comment by Coolthulhu** [comment]
I'll try debugging it and fiddling with the code later, but it will take a longer while since it takes 10 minutes just to link the debug build here.

**Comment by mugling** [comment]
Not disagreeing with you on that - just noting we are now targeting cygwin and that the buildbot doesn't cover this so it's all on to you.

Can you at least try those changes I put as I'm otherwise blind with this one

**Comment by Coolthulhu** [comment]
> Can you at least try those changes I put as I'm otherwise blind with this one

I tested with the moved constructor. That wasn't it - the actual problem was in static recipe in recipe dictionary file.
I'll try moving the static recipe inside functions, so that it is initialized only when needed. Maybe that will help.

**Comment by mugling** [comment]
Yes, an keep the change to `none` for the default recipe constructor

**Comment by mugling** [comment]
Any progress on this? I wasn't able to get cygwin working (too many compilation errors even with lots of local changes).

**Comment by Coolthulhu** [comment]
Recompiling with an attempt at fix

**Comment by Coolthulhu** [comment]
Didn't manage to get a non-segfaulting build yet. I'll probably need to review the code manually rather than relying on debugger.

**Comment by mugling** [comment]
Did you try both moving those static definitions into functions **and** setting the constructor parameter to `none` in the same build? I'm wary if it crashes in string_id.h` we have an existing bug I've managed to poke (although debugging may prove me wrong).

File: src/item_factory.cpp [hunk 16]
```diff
@@ -49,6 +49,8 @@ static t_string_set item_blacklist;
 static t_string_set item_whitelist;
 static bool item_whitelist_is_exclusive = false;
 
+static DynamicDataLoader::deferred_json deferred;
```
Comment (Coolthulhu) [inline src/item_factory.cpp]: This causes initialization of the json system before anything else is loaded. Even if it isn't a problem by itself, it could cause some of the problems with debugging.

**Comment by mugling** [comment]
I'm stuck where to go from here. I cant debug `cygwin`. Suggestions?

**Comment by Coolthulhu** [comment]
I have no idea honestly. I'll keep poking around, but it's hard to locate potential unspecified behavior or compiler bugs in a PR this big.

**Comment by mugling** [comment]
We only just began compiling for `cygwin` so it's likely to be brittle - we've effectively gone from not supporting it to first-class support. This does force development through a bottleneck. I'm going to try and work out a virtualised copy of windows and see if I can get a copy running so you're not working alone on this but that may take some time.

First step here could be to backup to the last version that did work for you (6387ea3) and work forward from there a few commits at a time.

**Comment by Coolthulhu** [comment]
OK, now this is some eldritch shit going on.
The crash happens when a completely unrelated line in options.cpp is invoked. Line not even modified in this PR.
It crashes at options.cpp:128, `while (std::getline(ssTemp, sItem, ',')) {`
It's probably Cygwin then.

I'll drop Cygwin for now, though it was useful for testing and debugging.

**Comment by Coolthulhu** [comment]
Cygwin-less `gcc` does work, testing now.

The sorting order is inverted. It's supposed to list hardest recipes on top.
Trivial fix.

It takes trillion years (2 seconds) to display the food tab here. While the build I'm testing it on is terribly unoptimized, this slowdown signifies an underlying problem with handling of tons of recipes.
The part where you compute `can_make_with_inventory` for ALL recipes on the list is a big problem. The old code did an implicit sort, only operating on recipes that have a possibility of being displayed. You calculate it for all just for the sorting. This is a badly scaling solution and for a sufficiently bloated set of content (say, core+more_everything_mod+blazemod, then opening a craft menu in a big craft base), the sort times would be unbearable AND they are recalculated on every full redraw.
Another problem is that you don't cache the value. It is the most expensive part of the calculation and so should be cached. Preferably without `mutable` hacks, like it is done for component availability, but with a more localized solution (wrapping recipes in structures).

Other than that it looks like a big improvement so far, but I may find more. I am naturally jinxed.

File: src/crafting.cpp [hunk 17]
```diff
@@ -1457,76 +1127,50 @@ bool player::disassemble( int dis_pos )
...
     } else if( activity.moves_left <= 0 ) {
-        activity.moves_left = recipe_time;
+        activity.moves_left = r.time;
     }
 
-    activity.values.push_back( dis_pos );
-    activity.coords.push_back( ground ? pos() : tripoint_min );
-    activity.str_values.push_back( recipe_ident );
+    activity.values.push_back( pos );
+    activity.coords.push_back( ground ? this->pos() : tripoint_min );
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Why the `this`?

File: src/crafting.cpp [hunk 18]
```diff
@@ -1457,76 +1127,50 @@ bool player::disassemble( int dis_pos )
...
+    const auto &r = recipe_dictionary::get_uncraft( obj.typeId() );
 
-    // No disassembly without proper light
-    // But book-ripping is OK
-    if( cur_recipe != nullptr &&
-        lighting_craft_speed_multiplier( *cur_recipe ) == 0.0f ) {
-        add_msg( m_info, _( "You can't see to craft!" ) );
-        activity.type = ACT_NULL;
+    // check sufficient light
+    if( lighting_craft_speed_multiplier( r ) == 0.0f ) {
```
Comment (Coolthulhu) [inline src/crafting.cpp]: `==` float comparisons are considered bad practice, would be better to use `<=`.

File: src/crafting.cpp [hunk 19]
```diff
@@ -1335,112 +1054,63 @@ void player::consume_tools( const std::vector<tool_comp> &tools, int batch,
...
 
-                have_all_qualities = false;
+    for( const auto &opts : dis.get_qualities() ) {
+        for( const auto &qual : opts ) {
+            if( !qual.has( inv ) ) {
+                if( alert ) {
+                    add_msg_if_player( m_info, _( "To disassemble %s, you need %s" ),
+                                       obj.tname().c_str(), qual.to_string().c_str() );
+                }
+                return false;
```
Comment (Coolthulhu) [inline src/crafting.cpp]: You return too early, meaning that the player will only get the first missing quality/tool on the list.

File: src/crafting.cpp [hunk 20]
```diff
@@ -387,22 +116,24 @@ void player::long_craft()
...
+        buffer << making.requirements().list_missing();
         popup( buffer.str(), PF_NONE );
         return false;
     }
 
-    return making->check_eligible_containers_for_crafting( batch_size );
+    return making.check_eligible_containers_for_crafting( batch_size );
 }
 
+recipe::recipe() : skill_used( NULL_ID ) {}
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Since some recipes (notably `null_recipe` in recipe_dictionary.cpp) are initialized in static context, this `NULL_ID` may cause problems depending on compiler. This dependence is known to cause builds for MacOSX to segfault at start.

File: data/json/recipes/recipe_weapon.json [hunk 21]
```diff
@@ -4792,7 +4792,7 @@
 },{
   "type" : "recipe",
   "result": "bokken",
-  "byproducts": { "id" : "splinter", "amount" : 30 },
+  "byproducts": [ [ "splinter" ] ],
```
Comment (Coolthulhu) [inline data/json/recipes/recipe_weapon.json]: Missing mult?

File: src/crafting.cpp (hunk 17 above)
Comment (mugling) [inline src/crafting.cpp]: Because `pos` is unfortunately already a parameter. I want to rip that section out in a future PR so leaving as-is for now.

File: src/crafting.cpp (hunk 20 above)
Comment (mugling) [inline src/crafting.cpp]: Fixed

File: src/crafting.cpp (hunk 19 above)
Comment (mugling) [inline src/crafting.cpp]: Want to redo qualities to actual support alternatives so will address this in another PR

File: data/json/recipes/recipe_weapon.json (hunk 21 above)
Comment (mugling) [inline data/json/recipes/recipe_weapon.json]: Fixed

**Comment by mugling** [comment]
@Coolthulhu implemented a caching solution for recipe availability

**Comment by Coolthulhu** [comment]
Better, but it's still significantly slower for a large set of recipes.
A good test is to fill a single tile fully with rags. This isn't even an overkill, as many crafting bases do have such a number of items.
Just spawn 4000 rags and wait for them to spill out (it takes a long while since tile volume is checked per every dropped rag), then open the crafting menu and switch tab to food.

**Comment by mugling** [comment]
@Coolthulhu suggestions?

**Comment by codemime** [comment]
IMHO the best solution would be revising the whole approach to stacking. It's totally insane to keep and handle 4000 indistinguishable rags (or whatever else represented as independent objects) instead of one object and the number.

**Comment by Coolthulhu** [comment]
Bringing some of the old approach would suffice. No need to reinvent stacking just for this PR.

Old approach did the sort implicitly: it went level by level, categorizing each recipe as craftable (at the moment) or uncraftable but visible. This approach allowed early exit as soon as all the necessary slots got filled.

**Comment by mugling** [comment]
The new version calculates availability of a recipe precisely once per invocation of the crafting GUI. How could this be more efficient?

**Comment by mugling** [comment]
The new version calculates availability only once per instance of the crafting GUI? How could it be more performant than that given you need to determine availability to apply correct sort ordering?

**Comment by codemime** [comment]
> Bringing some of the old approach would suffice. No need to reinvent stacking just for this PR.

Totally agree. I was talking about the whole project,  of course, not this particular PR (a side note, if you will). It certainly is solvable without "reinventing", but this performance issue is _one_ of the consequences of the original stacking design.

**Comment by mugling** [comment]
@codemime any suggestions as to this PR as I'm stuck for other obvious performance improvements (sans spaghetti-code)

File: src/crafting_gui.cpp [hunk 22]
```diff
@@ -202,9 +198,35 @@ const recipe *select_crafting_recipe( int &batch_size )
...
+                    return b->difficulty < a->difficulty;
+                } );
+
+                std::stable_sort( current.begin(), current.end(), [&]( const recipe * a, const recipe * b ) {
+                    return availability_cache[a] && !availability_cache[b];
+                } );
+
+                std::transform( current.begin(), current.end(),
+                std::back_inserter( available ), [&]( const recipe * e ) {
+                    return e->can_make_with_inventory( crafting_inv, helpers );
```
Comment (codemime) [inline src/crafting_gui.cpp]: > @codemime any suggestions as to this PR as I'm stuck for other obvious performance improvements (sans spaghetti-code)

You create the `availability_cache`, but, as far as I can say, never really use it (apart from sorting the `current` list). This line recalculates everything that was cached.

I'm gonna take a closer look at the algorithm. For now I'd suggest per-tab caching (to eliminate any recalculation after each tab was cached). Cache and display the first tab, meanwhile (while waiting for the input) cache the rest in the background.

Comment (mugling) [inline src/crafting_gui.cpp]: I see it now - that line needs to use the cache!

**Comment by mugling** [comment]
@Coolthulhu further improvements to performance from latest commit appears sufficient for now (sans rewrite)

File: src/crafting_gui.cpp [hunk 23]
```diff
@@ -202,9 +198,35 @@ const recipe *select_crafting_recipe( int &batch_size )
             if( batch ) {
                 batch_recipes( crafting_inv, helpers, current, available, chosen );
             } else {
-                // Set current to all recipes in the current tab; available are possible to make
-                pick_recipes( crafting_inv, helpers, current, available,
-                              tab.cur(), subtab.cur(), filterstring );
+
+                available.clear();
+                current.clear();
```
Comment (codemime) [inline src/crafting_gui.cpp]: Dupe of [these lines](https://github.com/CleverRaven/Cataclysm-DDA/pull/18286/files/967ce23bf96e41b7013fb2f48ddbbab19a0f95cc#diff-fc40e9036f0ff584805625e294aadbb0R196).

Comment (mugling) [inline src/crafting_gui.cpp]: Fixed

**Comment by mugling** [comment]
Have fixed all outstanding issues, would prefer merge now as is currently blocking all other PR's that affect recipes.

**Comment by Coolthulhu** [comment]
> How could it be more performant than that given you need to determine availability to apply correct sort ordering?

By stopping when you have enough craftable recipes.
You only ever need to determine the availability of enough recipes to fill the menu.

Though I guess it's acceptable, as the old version didn't check it either.

**Comment by Zireael07** [comment]
@Coolthulhu: sounds like a good idea for a next PR to avoid blocking other PRs.

**Comment by mugling** [comment]
@Coolthulhu noted. I'm probably going to rip out a lot of that code anyway in subsequent PR's so can attend to then. Crafting is one of the ugliest parts of the codebase to work on and it does also tend to block the PR queue.
